### ISSUE TITLE
Close a script in Editor when deleting it

### ISSUE THREAD

**Issue by me2beats** [godotengine/godot-proposals#1550]
**Describe the project you are working on:**
gdscript plugins

**Describe the problem or limitation you are having in your project:**
When deleting a script with godot filesystem (`right mouse click` - `Delete`), I also need to manually close it from script list in ScriptEditor (if it is opened). This can be inconvenient especially when doing it often.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Closing a script in Script Editor automatically when deleting it could resolve it.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
When deleting a script for example using `right mouse click` - `Delete` in Godot filesystem, the deleted script could be deleted from ScriptEditor script tabs automatically this way one wouldn't need to press `right click` - `Close` or `Ctrl+W` to close the script.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
using plugins - maybe, not sure about "a few lines" tho

**Is there a reason why this should be core and not an add-on in the asset library?:**
I think plugins can do that, but I alse think that at least having an opion to automatically close scripts when they are deleted would be useful at the core level.

**Comment by KoBeWi** [comment]
IMO this should be optional. VS Code keeps deleted scripts too (but appends `(deleted)` in their name) and it's useful sometimes.

**Comment by me2beats** [comment]
> VS Code keeps deleted scripts too (but appends `(deleted)` in their name) and it's useful sometimes.

In what cases? I think if I delete a script I'd like to forget about it

btw, if I delete several scripts at the same time, then in the scripts list I have to delete them one by one.
since it looks like there is no way to select multiple scripts at the same time in ScriptEditor

**Comment by TheDuriel** [comment]
_Definitely_ should be optional and off by default.

Scripts are intentionally not removed from the list so that you can recover them in case of a mistake.

**Comment by Calinou** [comment]
I think the better solution for now would be to eppend ` (deleted)` to the script name once it's removed from the FileSystem dock. Removing large amounts of scripts isn't something you do often after all.

**Comment by me2beats** [comment]
> Scripts are intentionally not removed from the list so that you can recover them in case of a mistake.

when deleting a script, Godot asks if you really need to delete the script
Imo this is enough to avoid doing mistakes.

**Comment by me2beats** [comment]
Btw: now scripts, when deleted, lose their names in the scripts list (you can see this after refreshing script list).

**Comment by pycbouh** [comment]
Yes, they probably lose it because there is no longer a file associated with the resource. I think the panel uses resource_path and thus deleted scripts have none. Some workaround is required to track obsolete paths, but I too think that we better keep them unless closed explicitly by a user.

**Comment by Ev1lbl0w** [comment]
I think a system like VSCode is good, but AFAIK it doesn't exactly work as described above; when you delete a file inside VSCode he **does** delete it entirely and closes it's open tab. It's only if something external to VSCode makes changes *(such as git commands)* that it keeps a reference to the old file *(the 'deleted' tag)*

So I think a similar system for Godot is better; if the user does something explicit inside Godot, then it shouldn't need to preserve the file *(and Godot already has a confirmation window for deleting a file, so keeping it afterwards is redundant)*; if a change to the filesystem happens externally, Godot can tag the open file as "deleted", similar to VSCode.

**Comment by Xrayez** [comment]
> Btw: now scripts, when deleted, lose their names in the scripts list (you can see this after refreshing script list).

> *and Godot already has a confirmation window for deleting a file, so keeping it afterwards is redundant*

Scripts are removed from the script list when deleting via the FileSystem docks since godotengine/godot#40647, because it literally says: "(no undo)". 🙂

**Comment by YuriSizov** [comment]
For the record, during a PR review meeting we've decided that closing scripts (and scenes) when you delete the underlying file is worse for usability. There is a related issue that needs to be addressed (https://github.com/godotengine/godot/issues/39469) and alongside it we can improve the presentation a bit.

We currently mark scripts as `[unsaved]` when they have no resource path (https://github.com/godotengine/godot/pull/40647), but we don't do the same for scenes. We've discussed that it may be better to add a special case where instead of "unsaved" scenes and scripts get marked as "deleted" in this situation.

**Comment by miv391** [comment]
How are you supposed to delete scenes? Only using external tools like vscode and/or File Explorer?

This is what happens when I delete a scene and a script.

1. select scene and script 
2. press Delete and confirm deletion 
3. select root scene (main.tscn in reproduction project of https://github.com/godotengine/godot/issues/100940) 
4. close Godot Editor 

=> Now I get two dialogs ("Please Confirm" dialog is below the "Files have been modified" dialog). Whatever buttons I click, the deleted scene is reincarnated and then I clean the mess using external tools.

![Image](https://github.com/user-attachments/assets/93de7f57-8785-4f3b-b61d-1427795d6252)

**Comment by kOyCHFqLJYDV** [comment]
The behavior I would expect to see is when actually deleting a scene (or, heck, any object), any closed editor tabs pertaining to that scene (or object) would also be closed. Since I opted to delete the scene. It was my intent to delete it, to remove it. There is no logical reason to physically delete a scene to then have it simply be re-created when saving the still open copy of it. Other than wanting to store it in a different location or something like that.

That functionality is covered by simply... renaming the scene or moving the scene in the file tree.

So, the current behavior feels a bit... unexpected to me personally.

In my personal workflow, I have come to work around the slight peculiarities by simply first closing all open tabs, then deleting the object I wish to delete from the file tree. This ensures I get the behavior I expect to see. It's a simple workaround but still a bit of a workaround for behavior that does not really seem logical to me.

If one deletes an object, one's intent is to have it be removed. So, why is it not removed outright?

**Comment by tetrapod00** [comment]
Regarding the specific set of reproduction steps you posted in this thread, pressing `Cancel` then `Don't save` cleanly deleted the scene for me. The files are still deleted even after re-opening the editor. So, at least I don't think there is bug there. Though I agree that in that particular case the UX is pretty bad.

**Comment by kOyCHFqLJYDV** [comment]
> Regarding the specific set of reproduction steps you posted in this thread, pressing `Cancel` then `Don't save` cleanly deleted the scene for me. The files are still deleted even after re-opening the editor. So, at least I don't think there is bug there. Though I agree that in that particular case the UX is pretty bad.

Well, it clearly isn't a bug per se. It is intended but unexpected behavior. It's bad UX indeed.
